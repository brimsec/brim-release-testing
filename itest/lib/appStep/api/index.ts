import {click, rightClick} from "./click"
import ingestFile from "./ingestFile"
import reload from "./reload"
import savePcap from "./savePcap"
import search from "./search"
import setSpan from "./setSpan"
import startApp from "./startApp"
import takeScreenshot from "./takeScreenshot"

export default {
  click,
  rightClick,
  ingestFile,
  reload,
  savePcap,
  search,
  setSpan,
  startApp,
  takeScreenshot
}
