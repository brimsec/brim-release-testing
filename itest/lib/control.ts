type PromiseFunc = () => Promise<any>

export function retry(f: PromiseFunc, attempts = 100, delay = 100) {
  return new Promise<any>((resolve, reject) => {
    f()
      .then((ret) => resolve(ret))
      .catch((err) => {
        setTimeout(() => {
          if (attempts === 1) {
            reject(err)
          } else {
            retry(f, attempts - 1, delay)
              .then((ret) => resolve(ret))
              .catch((err) => reject(err))
          }
        }, delay)
      })
  })
}

export const retryUntil = (
  f: PromiseFunc,
  cond_f: (arg0: any) => boolean,
  attempts = 15,
  delay = 1000
) =>
  // Retry f until a condition is met. cond_f receives the value from f after
  // it is resolved and is responsible for examining and returning a boolean to
  // determine if the condition is satisfied.
  retry(
    () =>
      new Promise((resolve, reject) => {
        f()
          .then((val) => {
            if (cond_f(val)) {
              resolve(val)
            } else {
              reject(new Error(`retryUntil condition failure: ${val}`))
            }
          })
          .catch((err) => {
            reject(`retryUntil promise failure: ${err}`)
          })
      }),
    attempts,
    delay
  )
