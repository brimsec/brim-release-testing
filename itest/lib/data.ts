export const dataSets = {
  // The purpose of this object is to have a single source of truth about
  // bounds and metrics related to test data. These numbers are also dependent
  // on product behavior. For example, if the default time window changes from
  // last 30 minutes to last hour, some of these numbers may become invalid.
  sample: {
    histogram: {
      rectAttrMin: 0,
      rectAttrMax: 1000,
      wholeSpaceDistinctPaths: 8,
      wholeSpaceRectsPerClass: 11
    },
    spaceName: "sample.pcap.brim"
  }
}
