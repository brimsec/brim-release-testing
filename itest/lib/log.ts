import path from "path"
import {createLogger, format, transports} from "winston"
import {itestDir} from "./env"

// When the flow-typed update produces an error....
// flow-typed/npm/winston_v3.x.x.js
// [1] 124â”‚     createLogger: <T>($winstonLoggerConfig<T>) => $winstonLogger<T>,

export const LOG = createLogger({
  level: "info",
  format: format.combine(
    format.timestamp({
      format: "YYYY-MM-DD HH:mm:ss.SSS"
    }),
    format.printf(({level, message, timestamp}) => {
      return `[${timestamp} ${level}]: ${message}`
    })
  ),
  transports: [
    new transports.Console({level: "info"}), // No idea how to make this runtime-configurable via command line. This has
    // mkdir -p semantics so the environment variable seems OK enough.
    new transports.File({
      filename: path.join(itestDir(), "itest.log"),
      level: "debug"
    })
  ]
})
