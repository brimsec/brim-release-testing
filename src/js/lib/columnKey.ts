import {Column} from "../types"

export default (col: Column) => {
  return `${col.name}:${col.type}`
}
