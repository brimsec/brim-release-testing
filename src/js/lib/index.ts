import {isEmpty} from "lodash"
import uniq from "lodash/uniq"

import {isArray} from "./is"
import animation from "./animation"
import date from "./date"
import doc from "./doc"
import file from "./file"
import obj from "./obj"
import open from "./open"
import transaction from "./transaction"
import win from "./win"

const array = {
  wrap(item: any) {
    if (isArray(item)) return item
    if (isEmpty(item)) return []
    return [item]
  }
}

export default {
  file,
  obj,
  doc,
  win,
  array,
  animation,
  open,
  date,
  transaction,
  compact: (array: any[]) => array.filter((item) => !!item),
  uniq,
  bounded: (num: number, [from, to]: [number, number]) => {
    return Math.max(from, Math.min(num, to))
  },
  move: <T>(array: T[], src: number, dest: number): T[] => {
    const copy = [...array]
    copy.splice(dest, 0, copy.splice(src, 1)[0])
    return copy
  },
  sleep: (ms: number) => new Promise<void>((r) => setTimeout(r, ms))
}

export type TimeUnit =
  | "years"
  | "year"
  | "y"
  | "months"
  | "month"
  | "M"
  | "weeks"
  | "week"
  | "w"
  | "days"
  | "day"
  | "d"
  | "hours"
  | "hour"
  | "h"
  | "minutes"
  | "minute"
  | "m"
  | "seconds"
  | "second"
  | "s"
  | "milliseconds"
  | "millisecond"
  | "ms"

export type TimeObj = {minutes: number; hours: number}

export type EpochObj = {sec: number; ns: number}
