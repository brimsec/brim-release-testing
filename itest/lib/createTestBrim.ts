import {Locator} from "../../src/js/test/createLocator"
import appStep from "./appStep/api"
// TODO in a future PR: remove direct logStep uses here.
import logStep from "./appStep/util/logStep"
import newAppInstance from "./newAppInstance"
import {retryUntil} from "./control"
import lib from "../../src/js/lib"

export default (name: string) => {
  let app
  let testIdx = 0

  beforeAll(async () => {
    app = await newAppInstance(name, ++testIdx)
    await appStep.startApp(app)
  })

  afterAll(async () => {
    if (app && app.isRunning()) await app.stop()
  })

  return {
    getApp() {
      return app
    },

    ingest(file: string) {
      return appStep.ingestFile(app, file)
    },

    mockSaveDialog(value: {
      canceled: boolean
      filePath: string | null | undefined
    }) {
      return app.mainProcess.emit(
        "spectron:mock",
        "windows:showSaveDialog",
        value
      )
    },

    setValue(locator: Locator, value: string) {
      return logStep(`set input ${locator.css} to ${value}`, () =>
        app.client
          .waitForVisible(locator.css)
          .then(() => app.client.setValue(locator.css, value))
      )
    },

    getText(locator: Locator) {
      return logStep(`get text from ${locator.css}`, async () => {
        await app.client.waitForVisible(locator.css)
        return await app.client.getText(locator.css)
      })
    },

    clickAppMenuItem(id: string) {
      return app.mainProcess.emit("spectron:clickAppMenuItem", id)
    },

    search(input: string) {
      return appStep.search(app, input)
    },

    click(locator: Locator) {
      return appStep.click(app, locator.css)
    },

    waitForText(locator: string, regex: RegExp) {
      return retryUntil(
        () => app.client.getText(locator),
        (s) => regex.test(s)
      )
    },

    waitUntil(fn: () => boolean, opts?: Object) {
      return app.client.waitUntil(fn, opts)
    },

    isVisible(locator: Locator) {
      return app.client.isVisible(locator.css)
    },

    isNotVisible(locator: Locator) {
      return app.client.isVisible(locator.css).then((v) => !v)
    },

    wait(ms: number) {
      return lib.sleep(ms)
    }
  }
}
