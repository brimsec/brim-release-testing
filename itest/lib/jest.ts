import {Application} from "spectron"

import appStep from "./appStep/api"
import {selectors} from "../../src/js/test/integration"
import {LOG} from "./log"

export const TestTimeout = 300000
// https://jestjs.io/docs/en/troubleshooting#unresolved-promises
// https://jestjs.io/docs/en/jest-object#jestsettimeouttimeout
jest.setTimeout(TestTimeout)

export const handleError = async (
  app: Application,
  initialError: Error,
  done: any
) => {
  appStep.takeScreenshot(app)
  let realError = undefined
  let notificationError = undefined
  LOG.error(`handleError: Test hit exception: ${initialError.message}`)
  LOG.info("handleError: Looking for any desktop app notifications")
  try {
    notificationError = await app.client.getHTML(selectors.notification, false)
  } catch (e) {
    notificationError = undefined
  }
  if (notificationError) {
    realError = new Error(
      "App notification '" +
        notificationError +
        "' (initial error: '" +
        initialError +
        "'"
    )
  } else {
    LOG.info("handleError: desktop app notification not found")
    realError = initialError
  }
  done.fail && done.fail(realError)
}

// Put tests in a wrapper that prints what test is being run. This is
// to eagerly print to the console where the run is so that hangs and other
// failures are easier to diagnose. Without this, Jest is pretty silent and
// it's hard to know what went wrong in a CI run. This MUST be paired with jest
// --verbose to work properly, otherwise LOG.info() calls seem to get
// buffered and printed later.
export const stdTest = (
  descr: string,
  f: (done: any) => void | Promise<void>,
  timeout: number = TestTimeout
) => {
  test(
    descr,
    (done) => {
      LOG.info(`Running test: ${descr}`)
      f(done)
    },
    timeout
  )
}
