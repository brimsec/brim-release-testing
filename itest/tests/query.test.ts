import {basename} from "path"

import {sprintf} from "sprintf-js"

import appStep from "../lib/appStep/api"
import newAppInstance from "../lib/newAppInstance"
import {handleError, stdTest} from "../lib/jest"

const simpleQueries = [
  "* | count()",
  "* | count() by _path | sort _path",
  "_path=conn | count()",
  "_path=conn | cut ts, id.orig_h, id.orig_p, id.resp_h, id.resp_p, proto | sort ts",
  "* | every 2s count() | sort ts",
  "* | every 2s count() by _path | sort ts, _path",
  "_path=x509 or _path=ssl | sort _path"
]

describe("Query tests", () => {
  let app
  let testIdx = 0

  beforeAll(async () => {
    app = newAppInstance(basename(__filename), ++testIdx)
    await appStep.startApp(app)
    await appStep.ingestFile(app, "sample.tsv")
    await appStep.setSpan(app, "Whole Space")
  })

  afterAll(async () => {
    if (app && app.isRunning()) {
      await app.stop()
    }
  })

  for (let i = 0; i < simpleQueries.length; i++) {
    const zql = simpleQueries[i]
    const testId = sprintf("%03d", i)
    stdTest(`query${testId}: "${zql}"`, (done) => {
      appStep
        .search(app, zql)
        .then((results) => {
          expect(results).toMatchSnapshot()
          done()
        })
        .catch((err) => {
          handleError(app, err, done)
        })
    })
  }
})
