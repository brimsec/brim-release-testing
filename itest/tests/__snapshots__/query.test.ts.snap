// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query tests query000: "* | count()" 1`] = `
Array [
  "count",
  "30",
]
`;

exports[`Query tests query001: "* | count() by _path | sort _path" 1`] = `
Array [
  "_path",
  "count",
  "capture_loss",
  "1",
  "conn",
  "19",
  "dns",
  "2",
  "files",
  "4",
  "ssl",
  "1",
  "stats",
  "1",
  "weird",
  "1",
  "x509",
  "1",
]
`;

exports[`Query tests query002: "_path=conn | count()" 1`] = `
Array [
  "count",
  "19",
]
`;

exports[`Query tests query003: "_path=conn | cut ts, id.orig_h, id.orig_p, id.resp_h, id.resp_p, proto | sort ts" 1`] = `
Array [
  "ts",
  "id.orig_h",
  "id.orig_p",
  "id.resp_h",
  "id.resp_p",
  "proto",
  "2020-02-25T16:03:05Z",
  "192.168.1.110",
  "56625",
  "172.217.9.142",
  "443",
  "tcp",
  "2020-02-25T16:03:06Z",
  "192.168.1.110",
  "57591",
  "172.217.9.142",
  "443",
  "tcp",
  "2020-02-25T16:03:07Z",
  "192.168.1.110",
  "55635",
  "18.246.31.137",
  "443",
  "tcp",
  "2020-02-25T16:03:07Z",
  "192.168.1.110",
  "55747",
  "13.52.5.22",
  "443",
  "tcp",
  "2020-02-25T16:03:08Z",
  "192.168.1.110",
  "55344",
  "52.37.243.173",
  "443",
  "tcp",
  "2020-02-25T16:03:08Z",
  "192.168.1.110",
  "55346",
  "52.37.243.173",
  "443",
  "tcp",
  "2020-02-25T16:03:08Z",
  "192.168.1.110",
  "57326",
  "173.194.201.189",
  "443",
  "tcp",
  "2020-02-25T16:03:08Z",
  "192.168.1.110",
  "57487",
  "192.30.253.125",
  "443",
  "tcp",
  "2020-02-25T16:03:08Z",
  "192.168.1.110",
  "55354",
  "52.37.243.173",
  "443",
  "tcp",
  "2020-02-25T16:03:09Z",
  "192.168.1.179",
  "47783",
  "192.168.1.255",
  "15600",
  "udp",
  "2020-02-25T16:03:10Z",
  "192.168.1.110",
  "57332",
  "64.233.179.189",
  "443",
  "tcp",
  "2020-02-25T16:03:10Z",
  "192.168.1.110",
  "55351",
  "18.205.93.211",
  "443",
  "tcp",
  "2020-02-25T16:03:11Z",
  "192.168.1.110",
  "54375",
  "192.168.1.254",
  "53",
  "udp",
  "2020-02-25T16:03:12Z",
  "192.168.1.110",
  "57635",
  "17.125.252.5",
  "443",
  "tcp",
  "2020-02-25T16:03:13Z",
  "192.168.1.110",
  "51848",
  "192.168.1.254",
  "53",
  "udp",
  "2020-02-25T16:03:13Z",
  "192.168.1.110",
  "57640",
  "209.216.230.240",
  "443",
  "tcp",
  "2020-02-25T16:03:15Z",
  "192.168.1.179",
  "51524",
  "192.168.1.255",
  "15600",
  "udp",
  "2020-02-25T16:03:15Z",
  "192.168.1.110",
  "57572",
  "172.217.6.138",
  "443",
  "tcp",
  "2020-02-25T16:03:15Z",
  "192.168.1.110",
  "57540",
  "172.217.1.138",
  "443",
  "tcp",
]
`;

exports[`Query tests query004: "* | every 2s count() | sort ts" 1`] = `
Array [
  "ts",
  "count",
  "2020-02-25T16:03:04Z",
  "2",
  "2020-02-25T16:03:06Z",
  "4",
  "2020-02-25T16:03:08Z",
  "6",
  "2020-02-25T16:03:10Z",
  "4",
  "2020-02-25T16:03:12Z",
  "4",
  "2020-02-25T16:03:14Z",
  "9",
  "2020-02-25T16:03:16Z",
  "1",
]
`;

exports[`Query tests query005: "* | every 2s count() by _path | sort ts, _path" 1`] = `
Array [
  "ts",
  "_path",
  "count",
  "2020-02-25T16:03:04Z",
  "conn",
  "1",
  "2020-02-25T16:03:04Z",
  "stats",
  "1",
  "2020-02-25T16:03:06Z",
  "conn",
  "3",
  "2020-02-25T16:03:06Z",
  "weird",
  "1",
  "2020-02-25T16:03:08Z",
  "conn",
  "6",
  "2020-02-25T16:03:10Z",
  "conn",
  "3",
  "2020-02-25T16:03:10Z",
  "dns",
  "1",
  "2020-02-25T16:03:12Z",
  "conn",
  "3",
  "2020-02-25T16:03:12Z",
  "dns",
  "1",
  "2020-02-25T16:03:14Z",
  "conn",
  "3",
  "2020-02-25T16:03:14Z",
  "files",
  "4",
  "2020-02-25T16:03:14Z",
  "ssl",
  "1",
  "2020-02-25T16:03:14Z",
  "x509",
  "1",
  "2020-02-25T16:03:16Z",
  "capture_loss",
  "1",
]
`;

exports[`Query tests query006: "_path=x509 or _path=ssl | sort _path" 1`] = `
Array [
  "2020-02-25T16:03:14Z",
  "ssl",
  "CeYm4A4XtVPVfF9wo6",
  "192.168.1.110",
  "57640",
  "209.216.230.240",
  "443",
  "TLSv12",
  "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384",
  "secp256r1",
  "news.ycombinator.com",
  "F",
  "⦻",
  "http/1.1",
  "T",
  "FXkLzR25cwYgeI5yI7",
  "Frbo5s1CRsV02JUoSe",
  "F1VUER3CattO8PiHWc",
  "CN=news.ycombinator.com,O=Y Combinator\\\\, Inc.,L=San Francisco,ST=California,C=US",
  "CN=DigiCert SHA2 Secure Server CA,O=DigiCert Inc,C=US",
  "⦻",
  "⦻",
  "ok",
  "2020-02-25T16:03:14Z",
  "x509",
  "FXkLzR25cwYgeI5yI7",
  "3",
  "074FE902C6B7D619884E1CA1854D9178",
  "CN=news.ycombinator.com,O=Y Combinator\\\\, Inc.,L=San Francisco,ST=California,C=US",
  "CN=DigiCert SHA2 Secure Server CA,O=DigiCert Inc,C=US",
  "2019-07-08T07:00:00Z",
  "2021-09-10T19:00:00Z",
  "rsaEncryption",
  "sha256WithRSAEncryption",
  "rsa",
  "2,048",
  "65537",
  "⦻",
  "news.ycombinator.com",
  "F",
  "⦻",
]
`;
